### PR TITLE
Remove ro service

### LINKED ISSUE #4567: Deprecate kubernetes-ro service
Now that the "secrets" type has been merged, and we have namespaces, I don't think it often makes sense to give blanket readonly access to objects.  So, lets deprecated it.

I don't know that we want to rip it out all in one go.
Nothing is using the `kubernetes-ro` service as far as I can see from searching the codebase, however some customers might depend on it, and maybe openshift does.  For GKE, I think we'd like it gone.  So, lets add a flag to the master that disables its creation.  If a customer needs it we can tell them how to get it back, but it will discourage new uses.
 
Once service-accounts land, we can give people that need it a way to automatically generate policy and credentials for things in pods.  Then we can delete `kubernetes-ro` it altogether.

### LINKED ISSUE #7963: The apiserver read-only API leaks bearer tokens
From a minion:
  curl http://kubernetes-master:7080/api/v1beta3/secrets
will return the bearer tokens for all service accounts.

I'm not sure if this is already taken care of as part of #5921.

### PR DESCRIPTION
**Pull request by lavalamp**
Fixes #4567
Fixes #7963

### CODE REVIEW

File: docs/accessing_the_api.md [hunk 1]
```diff
@@ -59,8 +54,7 @@ variety of uses cases:
 
 ## Expected changes
    - Policy will limit the actions kubelets can do via the authed port.
-   - Kube-proxy currently uses the readonly port to read services and endpoints,
-     but will eventually use the auth port.
+   - Kube-proxy uses the authenticated port.
```
Comment (erictune) [inline docs/accessing_the_api.md]: this change is complete.

**Comment by erictune** [comment]
LGTM, but title says WIP

File: docs/accessing_the_api.md (hunk 1 above)
Comment (lavalamp) [inline docs/accessing_the_api.md]: Cool, fixed.

**Comment by lavalamp** [comment]
I think I have e2e passing now.

Work items I still have to do before we merge this:
1. revisit my kube2sky change in this PR-- may not be necssary
2. Add a ClientFromKubernetesEnv function or something to the client package that'll automatically get you a client from the default service token.
3. Make kubectl do the same.

@erictune I'm thinking about punting on 3, there's no good way to test it yet. I want to run a registry on the cluster and push stuff to it and use pull secrets, but it'll require a lot of work in our build scripts to get it going. I don't know if we want to block on it. Unfortunately, this means there'd be no solution for things like the prometheus example, for which there's no obvious way to plug authentication in.

I suppose I could make it work and worry about testing later when we get registries in...

**Comment by lavalamp** [comment]
I just removed the kube2sky change; I don't think it's necessary anyway.

**Comment by roberthbailey** [comment]
> Add a ClientFromKubernetesEnv function ...

Just to be clear, this isn't going to suck a token in from the KubeEnv in the metadata server, but will grab it from a service account, right?

**Comment by lavalamp** [comment]
> Just to be clear, this isn't going to suck a token in from the KubeEnv in the metadata server, but will grab it from a service account, right?

Yes, it'll use the token from the service accounts, and later we'll make it also check the master's certs.

**Comment by lavalamp** [comment]
So the only thing left here is making kubectl work from a container. Still figuring out that long list of config merging thingies.

**Comment by lavalamp** [comment]
OK, I have kubectl running in a container. I'm going to run e2e again to make sure it (still) works.

**Comment by lavalamp** [comment]
Looks like heapster is broken with this, so I should fix that before we merge.

**Comment by lavalamp** [comment]
Heapster was not broken-- I got bit by the DNS problem. I've now got every e2e test working with this.

**Comment by lavalamp** [comment]
rebased

**Comment by lavalamp** [comment]
This should be ready to go now.

File: cluster/addons/fluentd-elasticsearch/es-image/elasticsearch_logging_discovery.go [hunk 2]
```diff
@@ -46,17 +41,7 @@ func main() {
...
-		glog.Fatalf("Error loading configuration from %s: %v", *kubeconfig, err.Error())
-	}
-
-	config, err := clientcmd.NewDefaultClientConfig(*settings, &clientcmd.ConfigOverrides{}).ClientConfig()
-	if err != nil {
-		glog.Fatalf("Failed to construct config: %v", err)
-	}
-
-	c, err := client.New(config)
+	c, err := client.NewInCluster()
```
Comment (erictune) [inline cluster/addons/fluentd-elasticsearch/es-image/elasticsearch_logging_discovery.go]: Fine for this PR, but I don't think we want to encourage a pattern where a program has to be in a cluster to run.  Should be some way to fall back to kubeconfig from a file, such as when doing ad-hoc testing.

File: cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -156,10 +154,6 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 		"will be used. If --bind-address is unspecified, the host's default interface will "+
 		"be used.")
 	fs.Var(&s.BindAddress, "public-address-override", "DEPRECATED: see --bind-address instead")
-	fs.IntVar(&s.ReadOnlyPort, "read-only-port", s.ReadOnlyPort, ""+
```
Comment (erictune) [inline cmd/kube-apiserver/app/server.go]: Removing ReadOnlyPort seems to me to be the most likely to break people's existing clusters.  Do you want to put it in a separate PR so it can be rolled back independent of the other changes?

File: cluster/addons/fluentd-elasticsearch/es-image/elasticsearch_logging_discovery.go (hunk 2 above)
Comment (lavalamp) [inline cluster/addons/fluentd-elasticsearch/es-image/elasticsearch_logging_discovery.go]: Yeah-- in another commit I have changes to kubectl code; people that use the same code path should be covered, but that is a very confusing bit of excessive overriding.

But this bit of code has essentially no usefulness outside of a cluster, so I think it's OK.

@satnam6502 you may want to glance at this.

**Comment by erictune** [comment]
LGTM

File: cluster/addons/fluentd-elasticsearch/es-image/elasticsearch_logging_discovery.go (hunk 2 above)
Comment (satnam6502) [inline cluster/addons/fluentd-elasticsearch/es-image/elasticsearch_logging_discovery.go]: Right, this code is designed to *only* run in a cluster. What's wrong with that? I thought this would be a not totally uncommon pattern?

File: cmd/kube-apiserver/app/server.go (hunk 3 above)
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Sure.

**Comment by lavalamp** [comment]
All commits except first moved to #9211.

**Comment by erictune** [comment]
still lgtm

**Comment by lavalamp** [comment]
Rebased & ready to go

**Comment by bgrant0607** [comment]
Waiting on Shippable.

**Comment by bgrant0607** [comment]
And e2e, if you'd like #8155 to not be rolled back in the case that this breaks e2e

**Comment by lavalamp** [comment]
Yes absolutely wait for e2e :)

**Comment by bgrant0607** [comment]
#8155 passed e2e, but I need to merge another rollback PR first.

**Comment by bgrant0607** [comment]
Sorry, I meant #9211 passed.

Tracking down another e2e failure at the moment - #9208.
